synapse/cores/sortedram.py

```diff
@@ -0,0 +1,222 @@
+import collections
+
+import sortedcontainers
+
+import synapse.cores.common
+from synapse.compat import isint
+
+
+class SortedRamCortex(synapse.cores.common.Cortex):
+    '''
+    SortedRamCortex is a RAM-backed Cortex that maintain int prop values
+     in sorted order to facilitate range queries.
+
+    Insertion performance appears to be a constant factor of 10% slower than
+     synapse.cores.ram.Cortex
+    Query by range performance is strictly better.
+    Query by range runtime is O(log n) vs O(n) for the ram:// cortex.
+    '''
+
+    def _initCortex(self):
+
+        def getValu(row):
+            return row[2]
+
+        def makePropSet():
+            return sortedcontainers.SortedSet(key=getValu)
+
+        self.rowsbyid = collections.defaultdict(set)
+        # rows by props with type(value) == string
+        self.srowsbyprop = collections.defaultdict(set)
+        # rows by props with type(value) == int
+        # split because sortedcontainers.SortedSet does not support keys with mixed types
+        self.irowsbyprop = collections.defaultdict(makePropSet)
+        self.rowsbyvalu = collections.defaultdict(set)
+
+        self.initSizeBy('ge',self._sizeByGe)
+        self.initRowsBy('ge',self._rowsByGe)
+
+        self.initSizeBy('le',self._sizeByLe)
+        self.initRowsBy('le',self._rowsByLe)
+
+        self.initSizeBy('range',self._sizeByRange)
+        self.initRowsBy('range',self._rowsByRange)
+
+    def _sizeByRange(self, prop, valu, limit=None):
+        min = None
+        if valu[0] is not None:
+            if not isint(valu[0]):
+                return 0
+            min = (None, None, valu[0], None)
+
+        max = None
+        if valu[1] is not None:
+            if not isint(valu[1]):
+                return 0
+            max = (None, None, valu[1], None)
+
+        rowsbyprop = self.irowsbyprop.get(prop)
+        if not rowsbyprop:
+            return 0
+
+        i = 0
+        for _ in rowsbyprop.irange(minimum=min, maximum=max, inclusive=(True, min and max)):
+            i += 1
+        return i
```

Comment:
under the hood it looks like they use a generator. so i don't think `len` will work. will test it and make the change if there's some magic that i'm missing.